bftengine/include/bftengine/ReplicaFactory.hpp

```diff
@@ -0,0 +1,73 @@
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "Replica.hpp"
+#include "ReplicaImp.hpp"
+#include "ReadOnlyReplica.hpp"
+#include "ReplicaLoader.hpp"
+
+namespace preprocessor {
+class PreProcessor;
+}
+
+namespace bftEngine {
+
+class ReplicaFactory {
+ public:
+  ReplicaFactory();
+  ~ReplicaFactory();
+  using IReplicaPtr = std::unique_ptr<IReplica>;
+  IReplicaPtr createNewReplica(const ReplicaConfig &,
+                               std::shared_ptr<IRequestsHandler>,
+                               IStateTransfer *,
+                               bft::communication::ICommunication *,
+                               MetadataStorage *,
+                               std::shared_ptr<concord::performance::PerformanceManager> pm,
+                               const std::shared_ptr<concord::secretsmanager::ISecretsManagerImpl> &sm,
+                               const std::function<void(bool)> &);
+  IReplicaPtr createNewReplica(const ReplicaConfig &,
+                               std::shared_ptr<IRequestsHandler>,
+                               IStateTransfer *,
+                               bft::communication::ICommunication *,
+                               MetadataStorage *,
+                               bool &erasedMetadata,
+                               std::shared_ptr<concord::performance::PerformanceManager> pm,
+                               const std::shared_ptr<concord::secretsmanager::ISecretsManagerImpl> &sm,
+                               const std::function<void(bool)> &);
+
+  IReplicaPtr createNewRoReplica(const ReplicaConfig &,
+                                 std::shared_ptr<IRequestsHandler>,
+                                 IStateTransfer *,
+                                 bft::communication::ICommunication *,
+                                 MetadataStorage *);
+
+  void setAggregator(std::shared_ptr<concordMetrics::Aggregator> aggregator);
+
+  static logging::Logger &logger() {
+    static logging::Logger logger_ = logging::getLogger("concord.replica_factory");
+    return logger_;
+  }
+
+ private:
+  void addNewPreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
+                          shared_ptr<IncomingMsgsStorage> &incomingMsgsStorage,
+                          shared_ptr<MsgHandlersRegistrator> &msgHandlersRegistrator,
+                          bftEngine::IRequestsHandler &requestsHandler,
+                          InternalReplicaApi &replica,
+                          concordUtil::Timers &timers,
+                          shared_ptr<concord::performance::PerformanceManager> &pm);
+
+  vector<std::unique_ptr<preprocessor::PreProcessor>> preProcessors_;
```

Comment:
The replica factory class does make sense (good idea to move these 3 functions), but what does it have to do with the pre-processor? pre-processor is not part of the replica. I would say it is part of the bft engine.
In addition, a factory should not (in general) hold any members varibles, unless it needs them to build objects.
Seems to me much better than before, but not yet the right solution on where to put the pre-processor (looks more like a forced solution).
I suggest to discuss this one on zoom.

Reply:
In addition, I understand you chose to keep the vector (so no need to fix the test). I think it's the right thing to do for now, although I totally don't agree that it should be kept like this, and it probably shouldn't be a vector, but we have a test which assumes that.

Reply:
The vector is used by those 3 ex-static functions, so in order to make that fly, I've put it in the class. If it is to be changed, there needs to be some logic rethinking, which I agree is significantly more difficult.